cupy/core/_fusion_analysis.py

```diff
@@ -0,0 +1,591 @@
+        """Generate new _FusionCudaArray object with a new memory space.
+        """
+        return self._generate_new_variable(_FusionCudaScalar, dtype, **kwargs)
+
+    def make_view(self, var, **kwargs):
+        assert isinstance(var, _FusionCudaArray)
+        serial_number = self._serial_number
+        self._serial_number += 1
+        ret = var.make_view(serial_number, **kwargs)
+        return self._normalize_variable(ret)
+
+    def broadcast_to(self, var, ashape, rshape):
+        """Make a view of the input array with the given shape.
+        """
+        return self.make_view(
+            var, ashape=ashape, rshape=rshape, broadcasted_from=var)
+
+    def rotate_with_axis(self, var, axis):
+        """Make a view of an array by rotating ``var`` with given axis.
+        """
+        assert isinstance(var, _FusionCudaArray)
+        return self.make_view(var, rotated_from=var, axis=axis)
+
+    def indexing(self, var, indices):
+        """Make a view of an array. by indexing ``var`` with given tuple.
+        """
+        skip = var.ndim - sum([isinstance(x, (int, slice)) for x in indices])
+        it = 0
+        ashape = []
+        rshape = []
+
+        if skip < 0:
+            raise IndexError('Too many indices for array.')
+
+        for index in indices:
+            if isinstance(index, int):
+                it += 1
+            elif isinstance(index, slice):
+                if not (index.start is None
+                        and index.stop is None
+                        and index.step in (1, -1, None)):
+                    raise NotImplementedError(
+                        'Only full range ``x[::]`` or reverse ``x[::-1]`` is '
+                        'supported for basic slicing in CuPy fusion.')
+                ashape.append(var.ashape[it])
+                rshape.append(var.rshape[it])
+                it += 1
+            elif index is None:
+                ashape.append(1)
+                rshape.append(1)
+            elif index is Ellipsis:
+                ashape.extend(var.ashape[it:it + skip])
+                rshape.extend(var.rshape[it:it + skip])
+                it += skip
+
+        ashape.extend(var.ashape[it:var.ndim])
+        rshape.extend(var.rshape[it:var.ndim])
+
+        return self.make_view(
+            var, indexed_from=var, index_key=indices,
+            ashape=tuple(ashape), rshape=tuple(rshape))
+
+    @property
+    def all_variables(self):
+        """Returns the list of all variables this class emitted.
+        """
+        return list(self._variables_dict.values())
+
+
+class _FusionHistory(object):
+    """History of operation exectuted in the target function of fusion.
+    """
+
+    def __init__(self, name):
+        self.name = name
+        self.vc = _VariableConductor()
+        self.shape_constraints = _fusion_shape._ShapeConstraints()
+        self.op_list = []
+        self.submodules = {}
+
+    @staticmethod
+    def _to_interface(x):
```

Comment:
How about `_make_interface`? "`self._to_interface`" might be confusing because it has an impression that `self` could be converted to an interface.